File: frontend/src/lib/degrade.ts
/**
 * Plain-language wording for a live-illustration failure.
 *
 * The app already degrades honestly when the live provider fails: it re-seals
 * the same reviewed scene with the offline placeholder, so the case still
 * completes and the manifest still records that the illustration is not a real
 * generative render. What a visitor saw, though, was a long wait and then a
 * grey box. Our sibling entry lived exactly that: the shared account ran out of
 * credit, the failure was written plainly to the logs, and the person on the
 * page had no way to know any of it.
 *
 * The server sends one token from a closed set (see `claimscene/degrade.py`).
 * This turns that token into a sentence, and nothing else: no status codes, no
 * upstream message, no account or service name. The operator's copy of the
 * failure stays in the logs, where the detail belongs.
 *
 * It matters more here than in a photo app. The illustration is the DISCLOSED
 * layer; the schematic carries the facts. So every message below is paired
 * with the same reassurance, because it is true: the schematic is drawn by
 * this codebase from the scene the human confirmed, and a failure at the
 * illustration provider cannot touch it.
 */

/** The honest fallback: used for `unknown`, for a missing kind, and for any
 *  token this build has not heard of. */
const FALLBACK = "The illustration service failed, and it did not say why.";

/** What the visitor is told, per failure kind. */
const CAUSE: Record<string, string> = {
  credit: "The illustration service was out of credit, so it would not run.",
  auth: "The illustration service did not accept our sign-in, so it would not run.",
  rate_limit: "The illustration service was handling too many requests and turned us away.",
  timeout: "The illustration service took too long to answer, so we stopped waiting.",
  unavailable: "The illustration service could not be reached.",
  unknown: FALLBACK,
};

/** The one-line cause, in words a claimant can read. An unrecognised kind
 *  (a server that learned a new one) falls back to the honest "we do not
 *  know" sentence rather than blanking the notice. */
export function degradeCause(kind: string | undefined | null): string {
  if (!kind) return FALLBACK;
  return CAUSE[kind] ?? FALLBACK;
}

/** What it means for this case. Constant on purpose: the consequence does not
 *  depend on which way the service failed, and the schematic is unaffected
 *  every single time. */
export const DEGRADE_CONSEQUENCE =
  "Your case is still sealed and complete. The schematic below is the factual layer, " +
  "drawn here from the scene you confirmed, and it is unaffected. Only the AI picture " +
  "is missing, and it was never evidence.";

/** Heading for the notice. */
export const DEGRADE_TITLE = "The AI picture could not be made this time";

File: src/claimscene/degrade.py
"""Classify a live media-provider failure into a small, closed vocabulary.

Why this module exists
----------------------
When the live illustration provider fails, ``claimscene.api._run_render``
already degrades honestly: it re-seals the SAME reviewed scene with the
offline provider, so the case still completes and the manifest still records
``illustration.degraded: true``. What the visitor saw, though, was only a long
wait followed by a placeholder — the *kind* of failure lived in Cloud Logging.
Our sibling entry Cinemory hit exactly that: the shared GMI account ran out of
credits, the backend logged ``GMICloud submit failed (402): Insufficient
credits``, and the person on the page had no way to know that from the app.

So this maps an exception to ONE of :data:`KINDS`. Two rules make it safe to
put that value in an HTTP response and in the sealed manifest:

1. **Only the kind crosses the wire.** The strings matched below come from
   upstream exception text, which can embed request URLs, model names, account
   identifiers, or a raw provider error body. None of that is returned — the
   caller gets a single lowercase token from a fixed set, and the frontend
   turns that token into a sentence a claimant can read.
2. **The real reason stays in the logs.** ``_run_render`` logs the full
   exception (``_log.exception``) with the kind alongside it, so an operator
   greps one line and gets both the plain-language category the visitor saw
   and the upstream detail they did not.

The matching is deliberately text-based rather than typed: the provider stack
(Genblaze → GMI Cloud → the underlying HTTP client) raises several unrelated
exception classes and folds the upstream status into the message. Walking the
``__cause__``/``__context__`` chain and looking for a status code or a
well-known phrase is what actually survives that. An unrecognised failure is
``"unknown"``, never a guess — the UI says "something went wrong at their end"
rather than inventing a cause.
"""
from __future__ import annotations

import re

#: The closed vocabulary. Every value here has a matching plain-language
#: sentence in the web client (``frontend/src/lib/degrade.ts``); adding a kind
#: means adding it there too, and the frontend falls back to the "unknown"
#: wording for anything it does not recognise.
KINDS = ("credit", "auth", "rate_limit", "timeout", "unavailable", "unknown")

#: How many links of the ``__cause__``/``__context__`` chain are inspected.
#: Bounded so a self-referencing or pathologically deep chain cannot spin.
_MAX_CHAIN = 8

# Ordered most-specific first: 402 must win over a generic "payment"/"auth"
# reading, and 429 over the 5xx/connection bucket.
_RULES: tuple[tuple[str, re.Pattern[str]], ...] = (
    ("credit", re.compile(
        r"\b402\b|insufficient\s+(credit|balance|fund)|out\s+of\s+credit"
        r"|payment\s+required|billing|top\s*[-_ ]?up|no\s+credits")),
    ("rate_limit", re.compile(r"\b429\b|rate[\s_-]?limit|too\s+many\s+requests|throttl")),
    ("auth", re.compile(
        r"\b40[13]\b|unauthori[sz]ed|forbidden|invalid\s+api[\s_-]?key"
        r"|authentication\s+fail|missing\s+api[\s_-]?key|api[\s_-]?key\s+not")),
    ("timeout", re.compile(r"timeout|timed\s+out|deadline\s+exceeded")),
    ("unavailable", re.compile(
        r"\b5\d\d\b|service\s+unavailable|bad\s+gateway|connection\s+(refused|reset|error)"
        r"|temporarily\s+unavailable|cannot\s+connect|name\s+resolution|unreachable")),
)


def _chain_text(exc: BaseException) -> str:
    """Lowercased ``ClassName: message`` for the exception and its causes.

    Both ``__cause__`` (explicit ``raise ... from``) and ``__context__``
    (implicit re-raise) are followed, because the provider stack uses both,
    and the status code is often only present on the innermost one. Cycles are
    impossible to follow twice: visited exceptions are tracked by identity.
    """
    parts: list[str] = []
    seen: set[int] = set()
    current: BaseException | None = exc
    while current is not None and len(parts) < _MAX_CHAIN and id(current) not in seen:
        seen.add(id(current))
        parts.append(f"{type(current).__name__}: {current}")
        current = current.__cause__ or current.__context__
    return " | ".join(parts).lower()


def classify(exc: BaseException) -> str:
    """One of :data:`KINDS` describing why the live provider could not deliver.

    Never raises: a classifier that blew up while handling a failure would
    turn an honest degrade into a 500, which is the exact opposite of what
    this codebase promises.
    """
    try:
        text = _chain_text(exc)
    except Exception:  # pragma: no cover - defensive; str() of a hostile repr
        return "unknown"
    for kind, pattern in _RULES:
        if pattern.search(text):
            return kind
    if isinstance(exc, TimeoutError):
        return "timeout"
    if isinstance(exc, (ConnectionError, OSError)):
        return "unavailable"
    return "unknown"
